backend/app/interactem/app/operators/registry.py

```diff
@@ -0,0 +1,104 @@
+import urllib.parse
+from typing import Any
+
+from httpx import AsyncClient
+from jsonpath_ng import parse
+
+GITHUB_API_URL = "https://api.github.com"
+
+
+class ContainerRegistry:
+    def __init__(
+        self,
+        url: str,
+        username: str,
+        password: str,
+    ):
+        self._url = url
+        self._username = username
+        self._password = password
+        self._token_ = None
+
+    async def __aenter__(self):
+        self._client = AsyncClient(base_url=self._url)
+
+        return self
+
+    async def __aexit__(self, exc_type, exc, tb):
+        await self._client.aclose()
+
+    async def _token(self) -> str:
+        if self._token_ is None:
+            params = {
+                "scope": "repository:*:pull",
+            }
+
+            r = await self._client.get(
+                "token", params=params, auth=(self._username, self._password)
+            )
+            r.raise_for_status()
+
+            self._token_ = r.json()["token"]
+
+        return self._token_
+
+    async def _headers(self):
+        token = await self._token()
+        return {"Authorization": f"Bearer {token}"}
+
+    async def _ghcr_images(self, org: str) -> list[str]:
+        url = f"{GITHUB_API_URL}/orgs/{org}/packages"
+        params = {"package_type": "container"}
+        headers = {
+            "Accept": "application/vnd.github.v3+json",
+            "Authorization": f"Bearer {self._password}",
+        }
+        r = await self._client.get(url, headers=headers, params=params)
+        r.raise_for_status()
+
+        path = parse("$[*].name")
+        matches = path.find(r.json())
+        if not matches:
+            raise Exception("Images not found")
+
+        return [f"{org}/{urllib.parse.quote(match.value)}" for match in matches]
+
+    async def images(self, namespace: str) -> list[str]:
+        # Special case for GitHub Container Registry, it currently doesn't support
+        # the _catalog endpoint
+        if "ghcr.io" in self._url:
+            return await self._ghcr_images(namespace)
+
+        headers = await self._headers()
+        r = await self._client.get("v2/_catalog", headers=headers)
+        r.raise_for_status()
+
+        return [f"{namespace}/{r}" for r in r.json()["repositories"]]
+
+    async def manifest(self, image: str, tag: str) -> dict[str, Any]:
+        headers = await self._headers()
+        headers["Accept"] = "application/vnd.oci.image.manifest.v1+json"
+
+        print(headers)
+
+        r = await self._client.get(f"v2/{image}/manifests/{tag}", headers=headers)
+        r.raise_for_status()
+
+        return r.json()
+
+    async def blob(self, image: str, digest: str) -> dict[str, Any]:
+        headers = await self._headers()
+        headers["Accept"] = "application/vnd.docker.distribution.manifest.v2+json"
```

Comment:
does this `"Accept"` field change ever? if so maybe a constant?

Reply:
sure, done